### DISCUSSION TITLE
Use fixture instance in non-test classes

### DISCUSSION THREAD

**Discussion by peppe-russo** [pytest-dev/pytest#11619] [Q&A]
Hello everyone, I'm writing a pytest plugin that uses [pytest-vnc](https://github.com/barneygale/pytest-vnc) as a dependency. The way I use it is to wrap it around my own custom fixture to override the `sleep` method, like so:
```
# conftest.py
@pytest.fixture(scope='session')
def qt_vnc(vnc):
    vnc.sleep = sleep
    return vnc
```
I then have an `Area` class in another file that I use quite often around my plugin, which needs to use that `VNC` instance (or more precisely, the `qt_vnc` fixture), example:
```
# area.py
from dataclasses import dataclass, asdict
from pytest_vnc import RectLike

@dataclass
class Area(RectLike):
    x: int
    y: int
    width: int
    height: int

    def capture():
        qt_vnc.capture(**asdict(self))  # <- How can access qt_vnc here?
```
The current way I'm doing this is to have an optional `vnc` field to the `Area` class and just pass the fixture to it when I instanciate the class, which doesn't look very elegant and it's not always possible as I use this class from places where I can't easily access the fixture.

Is there a "proper" way to make the `qt_vnc` fixture like a global singleton or something like that so I can access it from my custom classes? (The `Area` class is not the only class that needs to use it).

Thanks in advance,
Peppe

**Comment by portyu9** [comment]
Hi @peppe-russo 

A session-scoped fixture already gives every requester the same fixture instance, but pytest fixtures are **not globally importable singletons**.

The supported pattern is to keep the dependency explicit. If creating `Area` objects repeatedly is awkward, a factory fixture is a good way to hide that plumbing:

```python
from dataclasses import dataclass

import pytest


@dataclass
class Area:
    x: int
    y: int
    width: int
    height: int
    vnc: object

    def capture(self):
        return self.vnc.capture(
            x=self.x,
            y=self.y,
            width=self.width,
            height=self.height,
        )


@pytest.fixture
def area_factory(qt_vnc):
    def create(**kwargs):
        return Area(vnc=qt_vnc, **kwargs)

    return create
```

Then:

```python
def test_something(area_factory):
    area = area_factory(
        x=0,
        y=0,
        width=100,
        height=100,
    )
    area.capture()
```

Because `qt_vnc` is session-scoped, every `area_factory` request receives the same `qt_vnc` object.

`request.getfixturevalue()` exists for dynamically requesting fixtures from inside a pytest-managed test or fixture, but using it to make arbitrary classes pull dependencies out of pytest would effectively turn pytest into a service locator and tightly couple those classes to the test framework.

If many classes need VNC, another clean option is to create one fixture-backed service/context object containing the VNC dependency and have those objects receive that context.

Please mark this answer as accepted if it helped, thank you.
